### PR TITLE
feat: customize colors for types 

### LINKED ISSUE #47: Feature: Customize colors for types
Now we use default colors with Chrome API, it might be better to let user decide which color to use for a specified type.

Maybe we can add a color picker between the "type inputbar" and the "delete button"
image

### PR DESCRIPTION
**Pull request by flamintune**
Resolves #47

### CODE REVIEW

File: src/popup.tsx [hunk 1]
```diff
@@ -159,14 +193,60 @@ const Popup = () => {
...
+              ref={(node) => {
+                colorsRef.current[idx] = node;
+              }}
+              className={`bg-${colors[idx]}-500 w-4 h-4`}
+              style={{
+                backgroundColor: colors[idx],
+                flexShrink: 0,
+              }}
+            ></div>
+            <svg
```
Comment (MichaelYuhe) [inline src/popup.tsx:206]: I don't think using svg directly in the tsx file is a good idea. Can u try use the svg as an image instead?

File: src/popup.tsx [hunk 2]
```diff
@@ -159,14 +193,60 @@ const Popup = () => {
                 setTypes(newTypes);
               }}
             />
-
+            <div
+              ref={(node) => {
+                colorsRef.current[idx] = node;
+              }}
+              className={`bg-${colors[idx]}-500 w-4 h-4`}
+              style={{
```
Comment (MichaelYuhe) [inline src/popup.tsx:201]: className and style seems to be duplicated

File: src/popup.tsx [hunk 3]
```diff
@@ -147,6 +161,26 @@ const Popup = () => {
...
+                backgroundColor: colorOption,
+                borderColor: "transparent",
+                ["--tw-ring-color"]: colorOption,
+              }}
+              checked={color === colorOption}
+              value={colorOption}
+              onChange={(e) => {
+                setColor(e.target.value);
+                console.log(e.target.value);
```
Comment (MichaelYuhe) [inline src/popup.tsx:179]: remove console.log :D

File: src/popup.tsx [hunk 4]
```diff
@@ -147,6 +161,26 @@ const Popup = () => {
             </button>
           </div>
         </form>
+        <div className="flex items-center gap-x-2">
+          {DEFAULT_COLOR.map((colorOption) => (
+            <input
+              type="radio"
+              name="color"
+              className={`bg-${colorOption}-500 w-4 h-4 rounded-full`}
+              style={{
```
Comment (MichaelYuhe) [inline src/popup.tsx:170]: use className or style

File: src/popup.tsx [hunk 5]
```diff
@@ -159,14 +193,60 @@ const Popup = () => {
...
+              t="1702386306969"
+              class="icon"
+              viewBox="0 0 1024 1024"
+              version="1.1"
+              xmlns="http://www.w3.org/2000/svg"
+              p-id="1448"
+              width="16"
+              height="16"
+              onClick={(e) => {
+                let nextIdx =
```
Comment (MichaelYuhe) [inline src/popup.tsx:216]: const maybe better

File: src/popup.tsx [hunk 6]
```diff
@@ -159,14 +193,60 @@ const Popup = () => {
...
+                  (DEFAULT_COLOR.indexOf(colors[idx]) + 1) %
+                  DEFAULT_COLOR.length;
+                const newColors = [...colors];
+                newColors[idx] = DEFAULT_COLOR[nextIdx];
+                colorsRef.current[idx].animate(
```
Comment (MichaelYuhe) [inline src/popup.tsx:221]: it seems that you are using refs only for transition for colors? You can just add `transition-colors` in tailwind to do so. If there is any other reason using refs pls let me know~

File: src/utils.ts [hunk 7]
```diff
@@ -31,6 +31,17 @@ export const DEFAULT_GROUP = [
   "Utilities",
 ];
 
+export const DEFAULT_COLOR = [
```
Comment (MichaelYuhe) [inline src/utils.ts:34]: Since it's a fixed and unchanging choice for colors, maybe we can use enum for better types

**Comment by MichaelYuhe** [comment]
BTW, since it's a breaking change in popup(which some other users might don't like it, you can check my latest main branch, there is a feature flags tab in options page, you can decide whether to show the colors picker in popup

**Comment by flamintune** [comment]
> BTW, since it's a breaking change in popup(which some other users might don't like it, you can check my latest main branch, there is a feature flags tab in options page, you can decide whether to show the colors picker in popup

I have a question. If user have created serveral types without customizing colors, when he click this switch to enable customize colors, The color data for displaying types is sourced from where? And when he creates serveral types without customizing colors, do we need to generate a random color that is different from the previous type's color to save?

File: src/popup.tsx (hunk 6 above)
Comment (flamintune) [reply src/popup.tsx:221]: Yes, you're right. I was planning to utilize refs to create a translation animation with transform. The sole advantage of using refs might be the convenience it offers when altering the animation later on. Should I delete it...

File: src/popup.tsx (hunk 4 above)
Comment (flamintune) [reply src/popup.tsx:170]: ok,I'll use inline styles. Initially, I used className, but it had no effect, so I switched to style. I also noticed that Chrome's tab groups support 'grey' by default, but Tailwind CSS uses the American spelling 'gray'.

Comment (MichaelYuhe) [reply src/popup.tsx:170]: Maybe you can modify the `tailwind.config.js`

**Comment by MichaelYuhe** [comment]
> > BTW, since it's a breaking change in popup(which some other users might don't like it, you can check my latest main branch, there is a feature flags tab in options page, you can decide whether to show the colors picker in popup
> 
> I have a question. If user have created serveral types without customizing colors, when he click this switch to enable customize colors, The color data for displaying types is sourced from where? And when he creates serveral types without customizing colors, do we need to generate a random color that is different from the previous type's color to save?

By default we will use a random color unless user has specified a color for this type

**Comment by MichaelYuhe** [comment]
Sorry for the late merge, can you help me to solve the conflicts?

**Comment by flamintune** [comment]
It's okay, maybe I forgot to remind you. As a newbie, I really appreciate you letting me participate in the development of this feature and taking the time to review my low quality code and give feedback. If you did it yourself, you would be able to complete this feature very quickly. 😰😰

**Comment by MichaelYuhe** [comment]
> It's okay, maybe I forgot to remind you. As a newbie, I really appreciate you letting me participate in the development of this feature and taking the time to review my low quality code and give feedback. If you did it yourself, you would be able to complete this feature very quickly. 😰😰

Hey I really appreciate your work! Without your help I can't bring it to a next level myself!

**Review by MichaelYuhe — APPROVED** [review]

**Pull request merged**